### PR TITLE
[TT-3881] Implement the base for OIDC OAS conversion

### PR DESCRIPTION
**Pull request by furkansenharputlu**
This PR implements OAS conversion of OIDC authentication mode.

### CODE REVIEW

File: apidef/api_definitions.go [hunk 1]
```diff
@@ -868,6 +875,17 @@ func (a *APIDefinition) DecodeFromDB() {
 
 	makeCompatible("authToken")
 	makeCompatible("jwt")
+	// JWTScopeToPolicyMapping and JWTScopeClaimName are deprecated and following code ensures backward compatibility
+	a.JWTScopeClaim = ScopeClaim{
```
Comment (furkansenharputlu) [inline apidef/api_definitions.go:879]: I think you should do this only if `a.JWTScopeClaimName` is not empty but `ScopeClaimName` is empty.

File: apidef/oas/authentication.go [hunk 2]
```diff
@@ -122,6 +125,21 @@ func (a *Authentication) Fill(api apidef.APIDefinition) {
...
+		}
+
+		a.OIDC.Fill(api)
+	}
+
+	if ShouldOmit(a.OIDC) {
+		a.OIDC = nil
+	}
+
+	if a.GoPlugin == nil {
```
Comment (furkansenharputlu) [inline apidef/oas/authentication.go:140]: `GoPlugin`?

File: apidef/oas/authentication.go [hunk 3]
```diff
@@ -399,41 +421,6 @@ func (j *JWT) ExtractTo(api *apidef.APIDefinition) {
 	api.JWTExpiresAtValidationSkew = j.ExpiresAtValidationSkew
 }
 
-type Scopes struct {
```
Comment (furkansenharputlu) [inline apidef/oas/authentication.go:402]: I think this block should not be removed but should be updated.

File: apidef/oas/authentication.go [hunk 4]
```diff
@@ -623,6 +610,139 @@ func (h *HMAC) ExtractTo(api *apidef.APIDefinition) {
...
+	}
+}
+
+type Provider struct {
+	Issuer                string           `bson:"issuer,omitempty" json:"issuer,omitempty"`
+	ClientToPolicyMapping []ClientToPolicy `bson:"clientToPolicyMapping,omitempty" json:"clientToPolicyMapping,omitempty"`
+}
+
+type Scopes struct {
+	DefaultClaimName     string          `bson:"defaultClaimName,omitempty" json:"defaultClaimName,omitempty"`
```
Comment (furkansenharputlu) [inline apidef/oas/authentication.go:693]: Use the above struct, it is according to latest contract.

File: apidef/oas/authentication.go [hunk 5]
```diff
@@ -623,6 +610,139 @@ func (h *HMAC) ExtractTo(api *apidef.APIDefinition) {
...
+		api.JWTScopeToPolicyMapping = nil
+	}
+}
+
+type ClientToPolicy struct {
+	ClientID string `bson:"clientId,omitempty" json:"clientId,omitempty"`
+	PolicyID string `bson:"policyId,omitempty" json:"policyId,omitempty"`
+}
+
+type ClaimToPolicy struct {
```
Comment (furkansenharputlu) [inline apidef/oas/authentication.go:741]: This should be removed and `ScopeToPolicy` should be used.

File: apidef/oas/authentication_test.go [hunk 6]
```diff
@@ -55,18 +55,6 @@ func TestJWT(t *testing.T) {
 	assert.Equal(t, emptyJWT, resultJWT)
 }
 
-func TestScopes(t *testing.T) {
```
Comment (furkansenharputlu) [inline apidef/oas/authentication_test.go:58]: why do you remove this and add below?

File: apidef/oas/authentication.go [hunk 7]
```diff
@@ -404,11 +425,12 @@ type Scopes struct {
...
 }
 
-func (s *Scopes) Fill(api apidef.APIDefinition) {
-	s.ClaimName = api.JWTScopeClaimName
-
+func (s *Scopes) Fill(claimName string, scopeToPolicy map[string]string) {
 	s.ScopeToPolicyMapping = []ScopeToPolicy{}
-	for scope, policyID := range api.JWTScopeToPolicyMapping {
+
+	s.ClaimName = claimName
```
Comment (furkansenharputlu) [inline apidef/oas/authentication.go:431]: put this to first line

**Review by jeffy-mathew — APPROVED** [review]

File: apidef/oas/authentication.go [hunk 8]
```diff
@@ -417,15 +437,15 @@ func (s *Scopes) Fill(api apidef.APIDefinition) {
 	}
 }
 
-func (s *Scopes) ExtractTo(api *apidef.APIDefinition) {
-	api.JWTScopeClaimName = s.ClaimName
+func (s *Scopes) ExtractTo(scopeClaim *apidef.ScopeClaim) {
+	scopeClaim.ScopeClaimName = s.ClaimName
 
 	for _, v := range s.ScopeToPolicyMapping {
```
Comment (lghiur) [inline apidef/oas/authentication.go:443]: @furkansenharputlu  @jeffy-mathew  can this be transformed in more like a generic util func? ArrToMap or so?

**Review by lghiur — APPROVED** [review]

**Pull request merged**